### PR TITLE
rustdoc: remove def_ctor hack.

### PR DESCRIPTION
**Pull request by eddyb**
~~No longer necessary since we have `describe_def`.~~

Turns out `def_ctor` was used in conjunction with abusing `tcx.type_of(def_id)` working on both type definitions and `impl`s (specifically, of builtin types), but also reimplementing a lot of the logic that `Clean` already provides on `Ty` / `ty::TraitRef`.

The first commit now does the minimal refactor to keep it working, while the second commit contains the rest of the refactor I started (parts of which I'm not sure we need to keep).

### CODE REVIEW

File: src/librustdoc/passes/collect_trait_impls.rs [hunk 1]
```diff
@@ -67,8 +67,8 @@ pub fn collect_trait_impls(krate: Crate, cx: &DocContext<'_>) -> Crate {
         if !def_id.is_local() {
             inline::build_impl(cx, def_id, &mut new_items);
 
-            let auto_impls = get_auto_traits_with_def_id(cx, def_id);
-            let blanket_impls = get_blanket_impls_with_def_id(cx, def_id);
+            let auto_impls = get_auto_traits(cx, def_id, None);
+            let blanket_impls = get_blanket_impls(cx, def_id, None);
```
Comment (eddyb) [inline src/librustdoc/passes/collect_trait_impls.rs]: Hmm, this is technically wrong, `def_id` is not even pointing to a type, but rather an `impl`, which `get_def_from_def_id` *happens* to get the `Self` type for, using `type_of` - that, at least, explains the handling of primitive types (I was guessing type aliases, but no).

So I'll need to refactor things a bit more, turns out.

File: src/librustdoc/passes/collect_trait_impls.rs [hunk 2]
```diff
@@ -67,16 +67,14 @@ pub fn collect_trait_impls(krate: Crate, cx: &DocContext<'_>) -> Crate {
         if !def_id.is_local() {
             inline::build_impl(cx, def_id, &mut new_items);
 
-            let auto_impls = get_auto_traits_with_def_id(cx, def_id);
-            let blanket_impls = get_blanket_impls_with_def_id(cx, def_id);
-            let mut renderinfo = cx.renderinfo.borrow_mut();
+            // FIXME(eddyb) is this `doc(hidden)` check needed?
```
Comment (petrochenkov) [inline src/librustdoc/passes/collect_trait_impls.rs]: Well, looks like yes in theory, but not in practice because none of the "primitive" impls is hidden.

Comment (petrochenkov) [reply src/librustdoc/passes/collect_trait_impls.rs]: cc @QuietMisdreavus

File: src/librustdoc/passes/collect_trait_impls.rs [hunk 3]
```diff
@@ -155,14 +153,13 @@ impl<'a, 'tcx> SyntheticImplCollector<'a, 'tcx> {
...
         if i.is_struct() || i.is_enum() || i.is_union() {
-            if let (Some(hir_id), Some(name)) =
-                (self.cx.tcx.hir().as_local_hir_id(i.def_id), i.name.clone())
-            {
-                self.impls.extend(get_auto_traits_with_hir_id(self.cx, hir_id, name.clone()));
-                self.impls.extend(get_blanket_impls_with_hir_id(self.cx, hir_id, name));
-            } else {
-                self.impls.extend(get_auto_traits_with_def_id(self.cx, i.def_id));
-                self.impls.extend(get_blanket_impls_with_def_id(self.cx, i.def_id));
+            // FIXME(eddyb) is this `doc(hidden)` check needed?
```
Comment (petrochenkov) [inline src/librustdoc/passes/collect_trait_impls.rs]: My guess would be "not needed" or "needed only as optimization", if an item is hidden its impls won't be shown either way.

**Comment by petrochenkov** [comment]
Everything looks like a reasonable simplification, but the tests are failing.

**Comment by eddyb** [comment]
@petrochenkov I just fixed the tests (the output is now more in line with cross-crate `impl`s in general).

**Pull request merged**